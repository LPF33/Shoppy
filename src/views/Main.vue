<template>
  <section>
    <img :src="`${publicPath}img/wedding.JPG`" alt="wedding-photo" />
    <div v-if="schaetzchen" v-html="greeting" class="header"></div>
    <div v-else class="emoji">ðŸ¦‹</div>
    <aside>{{ wedding }} Tage bis zur Hochzeit</aside>
    <footer>
      <router-link :to="{ name: 'Budget' }" class="buttons">ðŸ’°</router-link>
      <router-link :to="{ name: 'Jambo' }" class="buttons">ðŸŽ²</router-link>
      <button @click="userLogout" class="buttons">ðŸ”’</button>
    </footer>
  </section>
</template>

<script>
export default {
  name: "Main",
  data() {
    return {
      publicPath: process.env.BASE_URL,
    };
  },
  methods: {
    userLogout() {
      this.$store.commit("unsubscribeFirebase");
      this.$store.commit("logout");
    },
  },
  computed: {
    wedding() {
      const weddingDate = new Date(2021, 8, 7);
      const now = new Date();
      return (
        Math.floor(
          (weddingDate.getTime() - now.getTime()) / (1000 * 3600 * 24)
        ) + 1
      );
    },
    schaetzchen() {
      return this.$store.state.user !== "lars@gmail.com";
    },
    greeting() {
      const now = new Date();
      const hours = now.getHours();
      if (5 <= hours && hours <= 11) {
        return "<h1>Guten Morgen ðŸŒž!</h1><h1>SchÃ¶nen Tag, meine Kleine â™¥</h1>";
      } else if (11 < hours && hours <= 14) {
        return "<h1>SchÃ¶nen Mittag ðŸŒ¼ðŸ¦‹</h1>";
      } else if (14 < hours && hours <= 17) {
        return "<h1>SchÃ¶nen Nachmittag â˜•</h1>";
      } else if (17 < hours && hours <= 22) {
        return "<h1>SchÃ¶nen Abend ðŸ¥°</h1><h1>mein kleines SchÃ¤tzchen ðŸ¥°</h1>";
      } else {
        return "<h1>Schlaf schÃ¶n ðŸ˜´</h1><h1> Gute Nacht, meine Kleine â™¥</h1>";
      }
    },
  },
};
</script>

<style scoped>
section {
  position: relative;
  background-color: #ffffff;
  height: 100%;
  color: #232323;
  font-weight: bold;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  z-index: 1;
}

img {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  max-width: 100vw;
  max-height: 100%;
  object-fit: contain;
  z-index: -1;
}

.header {
  padding: 20px;
}

aside {
  font-size: 1.5rem;
  padding: 20px;
  margin: auto 0;
  background-color: rgba(0, 0, 0, 0.507);
  border-radius: 10px;
  color: rgb(255, 255, 255);
}

footer {
  position: absolute;
  bottom: 10px;
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 0 10px;
}

.buttons {
  background-color: transparent;
  font-size: 3rem;
  height: 6rem;
  width: 6rem;
  border: 1px solid grey;
  border-radius: 50%;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  user-select: none;
  box-shadow: 0 0 8px black;
  text-decoration: none;
}

.emoji {
  font-size: 3rem;
  padding: 10px;
  margin-top: 1rem;
  user-select: none;
}
</style>
