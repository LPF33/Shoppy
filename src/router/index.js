import { createRouter, createWebHistory } from "vue-router";
import Main from "../views/Main.vue";
import Shoppy from "../views/Shoppy.vue";
import Jambo from "../views/Jambo.vue";
import store from "@/store";

const routes = [
  {
    path: "/",
    name: "Main",
    component: Main,
    meta: { requiresAuth: true },
  },
  {
    path: "/shoppy",
    name: "Shoppy",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: Shoppy,
  },
  {
    path: "/jambo",
    name: "Jambo",
    meta: { requiresAuth: true },
    component: Jambo,
  },
  {
    path: "/budget",
    name: "Budget",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Budget.vue"),
  },
  {
    path: "/login",
    name: "Login",
    meta: { guest: true },
    component: () => import("../views/Login.vue"),
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.requiresAuth)) {
    if (store.state.isAuthenticated) {
      next();
      return;
    }
    next("/login");
  } else {
    next();
  }
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.guest)) {
    if (store.state.isAuthenticated) {
      next("/");
      return;
    }
    next();
  } else {
    next();
  }
});

export default router;
