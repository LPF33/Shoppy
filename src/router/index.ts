import { createRouter, createWebHistory } from "vue-router";
import Main from "../views/Main.vue";
import Shoppy from "../views/Shoppy.vue";
import Jambo from "../views/Jambo.vue";
import Cooky from "../views/Cooky.vue";
import store from "@/store/index";

const routes = [
  {
    path: "/",
    name: "Main",
    component: Main,
    meta: { requiresAuth: true },
  },
  {
    path: "/shoppy",
    name: "Shoppy",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: Shoppy,
  },
  {
    path: "/jambo",
    name: "Jambo",
    meta: { requiresAuth: true },
    component: Jambo,
  },
  {
    path: "/budget",
    name: "Budget",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Budget.vue"),
  },
  {
    path: "/logout",
    name: "Logout",
    meta: { requiresAuth: true },
    // route level code-splitting
    // this generates a separate chunk (about.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: () => import("../views/Logout.vue"),
  },
  {
    path: "/cooky",
    name: "Cooky",
    meta: { requiresAuth: true },
    component: Cooky,
  },
  {
    path: "/login",
    name: "Login",
    meta: { guest: true },
    component: () => import("../views/Login.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    meta: { requiresAuth: true },
    component: () => import("../views/Settings.vue"),
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.matched.some((record) => record.meta.requiresAuth)) {
    if (store.state.isAuthenticated) {
      next();
      return;
    }
    next("/login");
  } else if (to.matched.some((record) => record.meta.guest)) {
    if (store.state.isAuthenticated) {
      next("/");
      return;
    }
    next();
  } else {
    next("/");
  }
});

export default router;
export { routes };
