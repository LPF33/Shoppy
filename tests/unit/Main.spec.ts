import { mount, flushPromises } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { routes } from "@/router/index";
import Main from "@/views/Main.vue";

let $store = {
  state: {
    user: "lars@gmail.com",
  },
  commit: jest.fn(),
};

beforeEach(() => {
  $store = {
    state: {
      user: "lars@gmail.com",
    },
    commit: jest.fn(),
  };
});

const router = createRouter({
  history: createWebHistory(),
  routes,
});

const mountMain = () => {
  return mount(Main, {
    data() {
      return {
        publicPath: "http://localhost:3000/",
      };
    },
    global: {
      plugins: [router],
      mocks: { $store },
    },
  });
};

describe("Testing the Main component", () => {
  it("Main component renders as expected when Lars is logged in", async () => {
    router.push("/");
    await router.isReady();
    const wrapper = mountMain();
    const mainWrapper = wrapper.find("section");
    expect(mainWrapper.exists()).toBeTruthy();
    expect(mainWrapper.find("img").exists()).toBeTruthy();
    expect(mainWrapper.find(".main-emoji").text()).toBe("ðŸ¦‹");
    expect(mainWrapper.findAll("h3")).toHaveLength(2);
  });

  it("Main component renders as expected when not Lars is logged in", async () => {
    $store.state.user = "not-lars@gmail.com";
    router.push("/");
    await router.isReady();
    const wrapper = mountMain();
    const mainWrapper = wrapper.find("section");
    expect(mainWrapper.exists()).toBeTruthy();
    expect(mainWrapper.find("img").exists()).toBeTruthy();
    expect(mainWrapper.find(".main-emoji").exists()).toBeTruthy();
    expect(mainWrapper.findAll("h3")).toHaveLength(2);
  });

  it("Main component renders as expected when not Lars is logged in", async () => {
    $store.state.user = "not-lars@gmail.com";
    router.push("/");
    await router.isReady();
    const wrapper = mountMain();
    const mainWrapper = wrapper.find("section");
    expect(mainWrapper.exists()).toBeTruthy();
    expect(mainWrapper.find("img").exists()).toBeTruthy();
    expect(mainWrapper.find(".main-emoji").exists()).toBeTruthy();
    expect(mainWrapper.findAll("h3")).toHaveLength(2);
  });

  /* it("Click on router-link with name Budget navigates to '/budget'", async () => {
    router.push("/");
    await router.isReady();
    const wrapper = mountMain();
    const jamboLink = wrapper.findAll("a")[0];
    expect(jamboLink.attributes("href")).toBe("/budget");
    jamboLink.trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe("/budget");
  });

  it("Click on router-link with name Jambo navigates to '/jambo'", async () => {
    router.push("/");
    await router.isReady();
    const wrapper = mountMain();
    const jamboLink = wrapper.findAll("a")[1];
    expect(jamboLink.attributes("href")).toBe("/jambo");
    jamboLink.trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe("/jambo");
  }); */
});
