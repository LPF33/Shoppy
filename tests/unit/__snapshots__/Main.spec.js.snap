// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Main component Main component renders as expected when Lars is logged in 1`] = `
<main><img src="http://localhost:3000/img/wedding.JPG">
  <div class="emoji">🦋</div>
  <aside>36 Tage bis zur Hochzeit</aside>
  <footer><a href="/budget" class="buttons">💰</a><a href="/jambo" class="buttons">🎲</a><button class="buttons">🔒</button></footer>
</main>
`;
